components/script/stylesheet_loader.rs

```diff
@@ -143,24 +146,32 @@ impl FetchResponseListener for StylesheetContext {
             let loader = StylesheetLoader::for_element(&elem);
             match self.source {
                 StylesheetContextSource::LinkElement { ref mut media, .. } => {
-                    let sheet =
-                        Arc::new(Stylesheet::from_bytes(&data, final_url,
-                                                        protocol_encoding_label,
-                                                        Some(environment_encoding),
-                                                        Origin::Author,
-                                                        media.take().unwrap(),
-                                                        Some(&loader),
-                                                        win.css_error_reporter(),
-                                                        ParserContextExtraData::default()));
-                    if elem.downcast::<HTMLLinkElement>().unwrap().is_alternate() {
-                        sheet.set_disabled(true);
+                    let link = elem.downcast::<HTMLLinkElement>().unwrap();
+                    // We must first check whether the generations of the context and the element match up,
+                    // else we risk applying the wrong stylesheet when responses come out-of-order.
+                    let should_continue = match self.request_generation_id {
+                        Some(gen) => gen == link.get_request_generation_id(),
+                        None      => true,
```

Comment:
nit: I don't think we use to indent fat arrows visually. In any case, you may be able to write this in a more compact way with `map_or`, as I did above.